File: src/main/resources/public/ts/services/DistributionService.ts
import {idiom, ng, notify} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {Distribution, Form} from '../models';

export interface DistributionService {
    list() : Promise<AxiosResponse>;
    listByResponder() : Promise<AxiosResponse>;
    listByForm(formId: number) : Promise<AxiosResponse>;
    listByFormAndStatus(formId: number, status: string, nbLines: number) : Promise<AxiosResponse>;
    listByFormAndResponder(formId: number) : Promise<AxiosResponse>;
    count(formId: number) : Promise<AxiosResponse>;
    get(distributionId: number) : Promise<AxiosResponse>;
    getByFormResponderAndStatus(formId: number) : Promise<AxiosResponse>;
    add(formId: number, distribution: Distribution) : Promise<AxiosResponse>;
    duplicateWithResponses(distributionId: number) : Promise<AxiosResponse>;
    update(distribution: Distribution) : Promise<AxiosResponse>;
    replace(distribution: Distribution) : Promise<AxiosResponse>;
    delete(distributionId: number) : Promise<AxiosResponse>;
}

export const distributionService: DistributionService = {
    async list() : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.list'));
            throw err;
        }
    },

    async listByResponder() : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/listMine`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.list'));
            throw err;
        }
    },

    async listByForm(formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/forms/${formId}/list`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.list'));
            throw err;
        }
    },

    async listByFormAndStatus(formId: number, status: string, nbLines: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/forms/${formId}/list/${status}?nbLines=${nbLines}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.list'));
            throw err;
        }
    },

    async listByFormAndResponder(formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/forms/${formId}/listMine`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.list'));
            throw err;
        }
    },

    async count(formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/forms/${formId}/count`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.count'));
            throw err;
        }
    },

    async get(distributionId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/${distributionId}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.get'));
            throw err;
        }
    },

    async getByFormResponderAndStatus(formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/distributions/forms/${formId}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.get'));
            throw err;
        }
    },

    async add(formId: number, distribution: Distribution) : Promise<AxiosResponse> {
        try {
            return http.post(`/formulaire/distributions/forms/${formId}/add`, distribution);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.create'));
            throw err;
        }
    },

    async duplicateWithResponses(distributionId: number) : Promise<AxiosResponse> {
        try {
            return http.post(`/formulaire/distributions/${distributionId}/duplicate`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.create'));
            throw err;
        }
    },

    async update(distribution: Distribution) : Promise<AxiosResponse> {
        try {
            return http.put(`/formulaire/distributions/${distribution.id}`, distribution);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.update'));
            throw err;
        }
    },

    async replace(distribution: Distribution) : Promise<AxiosResponse> {
        try {
            return http.delete(`/formulaire/distributions/${distribution.id}/replace/${distribution.original_id}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.update'));
            throw err;
        }
    },

    async delete(distributionId: number) : Promise<AxiosResponse> {
        try {
            return http.delete(`/formulaire/distributions/${distributionId}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.distributionService.delete'));
            throw err;
        }
    }
};

export const DistributionService = ng.service('DistributionService',(): DistributionService => distributionService);

File: formulaire/src/main/java/fr/openent/formulaire/service/DistributionService.java
package fr.openent.formulaire.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

import java.util.List;

public interface DistributionService {
    /**
     * List all the distributions of the forms sent by me
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void listBySender(UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the distributions of the forms sent to me
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void listByResponder(UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the distributions of a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void listByForm(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the distributions of specific forms
     * @param formIds form identifiers
     */
    Future<JsonArray> listByForms(JsonArray formIds);

    /**
     * List all the distributions for a specific form sent to me
     * @param formId form identifier
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void listByFormAndResponder(String formId, UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the distributions of a specific form with specific status
     * @param formId form identifier
     * @param status status
     * @param nbLines number of lines already loaded
     * @param handler function handler returning JsonArray data
     */
    void listByFormAndStatus(String formId, String status, String nbLines, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the distributions of a specific question of a specific form with specific status
     * @param formId form identifier
     * @param status status
     * @param questionId question identifier
     * @param nbLines number of lines already loaded
     * @param handler function handler returning JsonArray data
     */
    void listByFormAndStatusAndQuestion(String formId, String status, String questionId, String nbLines, Handler<Either<String, JsonArray>> handler);

    /**
     * Get the number of distributions for a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonObject data
     */
    void countFinished(String formId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get the number of distributions with TO_DO status for a specific form
     * @param formId form identifier
     * @param user user connected
     */
    Future<JsonObject> countMyToDo(String formId, UserInfos user);

    /**
     * Get the number of distributions with FINISHED status for a specific form
     * @param formId form identifier
     * @param user user connected
     */
    Future<JsonObject> countMyFinished(String formId, UserInfos user);

    /**
     * Get a specific distribution by id
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String distributionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get a specific distribution by form, responder and status
     * @param formId form identifier
     * @param user user connected
     * @param handler function handler returning JsonObject data
     */
    void getByFormResponderAndStatus(String formId, UserInfos user, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a new distribution based on an already existing one
     * @param distribution JsonObject data
     */
    Future<JsonObject> add(JsonObject distribution);

    /**
     * Duplicate a distribution by id
     * @param distributionId distribution identifier
     */
    Future<JsonObject> duplicateWithResponses(String distributionId);

    /**
     * Update a specific distribution
     * @param distributionId distribution identifier
     * @param distribution JsonObject data
     * @param handler function handler returning JsonObject data
     */
    void update(String distributionId, JsonObject distribution, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete a specific distribution
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonObject data
     */
    void delete(String distributionId, Handler<Either<String, JsonObject>> handler);

    /**
     *
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void getResponders(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Distribute a form to a list of responders who do not already have respond right
     * @param formId form identifier
     * @param user user connected
     * @param newResponders responders to whom we give respond right
     * @param handler function handler returning JsonObject data
     */
    void createMultiple(String formId, UserInfos user, List<JsonObject> newResponders, Handler<Either<String, JsonObject>> handler);

    /**
     * Update active status of the wanted distributions for a specific form
     * @param active status value to give to all the distributions
     * @param formId form identifier
     * @param responder_ids ids of the responders of the distributions to update
     * @param handler function handler returning JsonObject data
     */
    void setActiveValue(boolean active, String formId, List<String> responder_ids, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete distributions too old according to lifetime RGPD chosen
     * @param handler function handler returning JsonArray data
     */
    void deleteOldDistributions(Handler<Either<String, JsonArray>> handler);
}